### ISSUE TITLE
Show warning messages for commands in command log

### ISSUE THREAD

**Issue by Bkucera** [cypress-io/cypress#4917]
It would be useful for a command to print a warning instead of an error in the command log.

This would allow commands to alert the user something they have done may lead to unexpected behavior, such as pressing the `{home}` key while we cannot preform the default action of scrolling.

**Comment by flotwig** [comment]
A bunch of warnings already exist in the driver, though they just dump to `console.warn`, where most users won't notice it:

https://github.com/cypress-io/cypress/blob/0cbd95460199176896d7cb935df73777b0e878d7/packages/driver/src/cy/commands/xhr.coffee#L418-L418

https://github.com/cypress-io/cypress/blob/0cbd95460199176896d7cb935df73777b0e878d7/packages/driver/src/cy/commands/request.coffee#L100-L100

https://github.com/cypress-io/cypress/blob/0cbd95460199176896d7cb935df73777b0e878d7/packages/driver/src/cy/commands/agents.coffee#L234-L234

https://github.com/cypress-io/cypress/blob/0cbd95460199176896d7cb935df73777b0e878d7/packages/driver/src/cypress/cookies.coffee#L48-L58

https://github.com/cypress-io/cypress/blob/0cbd95460199176896d7cb935df73777b0e878d7/packages/driver/src/cypress/cy.coffee#L65-L67

https://github.com/cypress-io/cypress/blob/0cbd95460199176896d7cb935df73777b0e878d7/packages/driver/src/cypress/server.coffee#L45-L51

https://github.com/cypress-io/cypress/blob/0cbd95460199176896d7cb935df73777b0e878d7/packages/driver/src/cypress/xml_http_request.coffee#L127-L137

**Comment by jennifer-shehane** [comment]
Yeah, we mostly use console.warn for deprecations currently.

I like this idea though. I feel like this is what most people would want for the 'uncaught:exception' messages we give. They do not want a test failure, but a warning may be nice. 

Relevant to error improvements: https://github.com/cypress-io/cypress/issues/3762

**Comment by jennifer-shehane** [comment]
Played around with some of the design of this warning.

<img width="1792" alt="error-design-warnings" src="https://user-images.githubusercontent.com/1271364/62519355-6487e480-b851-11e9-8856-d09bd300827d.png">

**Comment by jennifer-shehane** [comment]
Second iteration for design:

<img width="1792" alt="error-design-warning-2" src="https://user-images.githubusercontent.com/1271364/62595227-8f327580-b903-11e9-9cc7-ebdea2f16155.png">

**Comment by Bkucera** [comment]
@jennifer-shehane I like the yellow background with dark yellow/orange text :+1:

**Comment by jennifer-shehane** [comment]
- Establish pattern that draws the user's attention to something that *functionally* is fine - does not cause a failure, but that they should be aware of.
- Maybe have warn `types` - like `deprecation`, `antipattern`, `misuse` - even when it's not associated with a command?

**Example: unnecessary assertion**

```js
cy.get('button').should('exist')
```

While functionally this won't break - the .should('exist') is unnecessary.

**Example: antipattern?**

```js
cy.visit(...)
cy.route(...) // definition for a route that has already happened
```

You likely don't want to define the listener for the cy.route() after cy.visit() in this situation. 

**Example: deprecation warnings**

We do these in the console.warn now, but why not have it in the Command Log?

**Example: async weirdness/mixing Promises and Cypress**

**Example: uncaught exceptions**

### Implementation:

```js
Cypress.log({
  warning: ...
})
```

Should be expandable, but not expanded by default - unless outside of Interactive Mode.

Have ability to turn off warnings - maybe by type of warning.

- Possibly look at eslint / linting libraries for API inspiration (config verbiage). 
- Also look at BlueBird config. http://bluebirdjs.com/docs/api/promise.config.html

```js
Cypress.Warnings.defaults({
  warnOnMixedAsyncUsage: false,
  warnOnUncaughtExceptions: false,
})
```

#### Warning *event* (aka. like URL, XHR events...)

For uncaught exceptions - we do some complex things to associated the uncaught exception with a test / command - we should just create an uncaught exception *event* and highlight that also in red. 

**Side discussion** As a breaking change - should switch logic to not fail by default on uncaught:exception - make them turn on failure for uncaught exception

Maybe add a code frame for the error events - honestly we should just design the UI of the warnings to be exactly like the 'failed' errors - with a 'docsUrl', 'code frame', 'stack trace', etc.

#### Dashboard

We could pin warnings to Dashboard results on the specs, "warnings tab?" - code frames - however Dashboard wants to design. 

Could be multiple warnings per test - would require API changes.

**Comment by jennifer-shehane** [comment]
Another application for displaying a 'warning' instead of throwing. A person could screenshot an element that is 0 width and/or 0 height as described here: https://github.com/cypress-io/cypress/issues/5149

I wouldn't necessarily fail the test in this case but would want to show a warning that there will be no screenshot taken.

**Comment by cmcnicholas** [comment]
The ability to warn and have these not fail a test run but be visible on cypress dashboard would be fantastic, we have several use cases really only specific to our application where we would like to stub tests in preparation for QA in situations where the amount of tests to write is a huge effort.

**Comment by YoungElPaso** [comment]
> The ability to warn and have these not fail a test run but be visible on cypress dashboard would be fantastic, we have several use cases really only specific to our application where we would like to stub tests in preparation for QA in situations where the amount of tests to write is a huge effort.

This is exactly my use case as well.  I have a complex mechanism of filtering 'dev' tests vs 'QA' tests so that no test I've written fails against the environment it is filtered for but this is not great for non-developers.

I want to be able to write tests useful for our QA process that aren't yet satisfied but *should* be.  I'll grant, this is a workflow/organization problem, but being able to illustrate to stakeholders that we can write tests that will eventually all pass is a big selling point for using testing like Cypress. Conversely having those tests just 'fail' right away (cause the code isn't yet developed) doesn't enamor the non-technical people to the process. They just see all the red.  It's also bad for a CI pipeline situation (although my filtering mechanism is well suited to that).

Anyway, a 'warning' would be useful for some situations although I recognize the potential for abuse (everything becomes a warning).

**Comment by Cleudi** [comment]
I also see the need for `cy.warn`. We maintain an extended cypress custom command library which is used by other project. No one will ever notice eg. deprecation notes in the console log. In contrast, having (orange) warnings in the command log will be noticed for sure (as long as the functionality is not abused). 

Maybe useful as well (or instead of `cy.warn`). Allow passing a custom color to `cy.log` - so there can be visual distinction between different use cases (eg. deprecation vs. anti-pattern vs. ... ).

**Comment by aleitch1** [comment]
Yeah, I could REALLY use a case for passing warnings or annotations to students about the tests they're trying to pass! All green or gray means they just don't read the messages.

**Comment by verheyenkoen** [comment]
For deprecations I would also find a badge in the Dashboard useful, like you do for flaky tests.

![image](https://user-images.githubusercontent.com/599538/141761351-6e0208a9-95a6-4916-bb11-bd687984988f.png)

We have some projects without active development where tests are just ran in CI daily. Whenever I get a Cypress release note I just upgrade them in a script (I should test locally first, I know, but I don't care if they fail first). There hardly ever is a problem that I didn't foresee by reading the release notes, even for major releases. I do look at the dashboard sometimes and investigate flaky tests. Would be good to be able to catch deprecations in the warning stage though.

**Comment by chrispepper1989** [comment]
This would be fantastic for the new access accessibility tests I want to bring in. Having tests fail at the moment would be a nightmare as there is a long list of legacy issues. But warnings would put us on the right track to fix them!

**Comment by SunshineyDay** [comment]
Another good use case for cy.warn(). We have a great suite of Cypress tests that we run as a smoke test whenever we do a new release of our app. These tests run against the live API and database rather than interceptors with fixtures. If the tests are executed against an api + database environment that does not yet contain some of the datasets required for the tests to pass it would be useful to be able to skip those tests and output a cy.warn() rather than having the test fail because of the lack of data. I agree with a comment earlier in this thread that it would also be useful to surface those warnings alongside passing and failing tests in the log and have warnings counted in addition to passes and fails at the top of the test execution window.

**Comment by furkanmustafa** [comment]
Would be nice to distinguish "passed" tests from "passed with warnings".
